subsys/net/lib/mqtt_sock/mqtt.c

```diff
@@ -121,54 +122,17 @@ static int client_connect(struct mqtt_client *client)
 
 static int client_read(struct mqtt_client *client)
 {
-	u8_t *buf = client->rx_buf + client->internal.rx_buf_datalen;
-	u32_t data_len = client->rx_buf_size - client->internal.rx_buf_datalen;
-	int processed;
 	int err_code;
 
-	err_code = mqtt_transport_read(client, buf, &data_len);
-	if (err_code == -EAGAIN) {
-		/* No data received. Try again later. */
-		return 0;
+	if (client->internal.remaining_payload > 0) {
+		return -EPERM;
```

Comment:
This can't be the right error, can it?

Reply:
Well, notionally, this is "state [transition] error". Would need to go thru errno's and see what's the closest to it - maybe nothing, POSIX errnos suck :-(

EINPROGRESS is probably more close to sockets, and thus confusing (is that *socket* error that we got or *mqtt lib* error?). EBUSY sounds good then, yeah.

Reply:
Just realized why it came up to me to use `EPERM` here, beacuse it doesn't make much sense basing on Zephyr's `errno.h`. In Zephyr we have:
```
EPERM 1		/* Not owner */
```
While errno man pages state:
```
EPERM           Operation not permitted (POSIX.1-2001).
```
IMO the two descriptions are completely different, and the latter one makes this choice a little more reasonable, as `mqtt_input` shall not be called when payload is available to read. Anyway, if there's a better choice, let's change it.

Reply:
Well, EPERM is "Operation not permitted", regardless of what Zephyr's bugs in the description say. And "not permitted" in the sense of "permission", i.e. access control, as the lexical form of EPERM clearly says. Regardless of any attempts to clarify its meaning/associate additional constraints, people will keep thinking of it as "there's not enough permission to do that".

For a classical reference (rant) on that matter, see http://blog.unclesniper.org/archives/2-Linux-programmers,-learn-the-difference-between-EACCES-and-EPERM-already!.html - calls like that are futile, things should be named right, to be used right.

(Note that explanation there slightly favors your usage - but not really, from my reading of it. It says EPERM should be used based on the static property of the system, like hardlinks to directory are forever not permitted, whereas we here talk about dynamic property, i.e. "in the current state, that operation is not permitted"). Again, it's sad that POSIX doesn't provide generic "invalid state" error :-(.